subscription_manager.go

```diff
@@ -0,0 +1,271 @@
+package river
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/riverqueue/river/internal/baseservice"
+	"github.com/riverqueue/river/internal/jobcompleter"
+	"github.com/riverqueue/river/internal/jobstats"
+	"github.com/riverqueue/river/internal/util/sliceutil"
+	"github.com/riverqueue/river/rivertype"
+)
+
+type subscriptionManager struct {
+	baseservice.BaseService
+
+	subscribeCh <-chan []jobcompleter.CompleterJobUpdated
+	wg          sync.WaitGroup
+
+	statsMu        sync.Mutex // protects stats fields
+	statsAggregate jobstats.JobStatistics
+	statsNumJobs   int
+
+	mu               sync.Mutex // protects subscription fields
+	subscriptions    map[int]*eventSubscription
+	subscriptionsSeq int // used for generating simple IDs
+}
+
+func newSubscriptionManager(archetype *baseservice.Archetype, subscribeCh <-chan []jobcompleter.CompleterJobUpdated) *subscriptionManager {
+	return baseservice.Init(archetype, &subscriptionManager{
+		subscribeCh:   subscribeCh,
+		subscriptions: make(map[int]*eventSubscription),
+	})
+}
+
+// ResetSubscribeChan is used to change the channel that the subscription
+// manager listens on. It must only be called when the subscription manager is
+// stopped.
+func (sm *subscriptionManager) ResetSubscribeChan(subscribeCh <-chan []jobcompleter.CompleterJobUpdated) {
+	sm.subscribeCh = subscribeCh
+}
+
+func (sm *subscriptionManager) Start(ctx context.Context) error {
+	sm.wg.Add(1)
+	go func() {
+		defer sm.wg.Done()
+
+		for updates := range sm.subscribeCh {
+			sm.distributeJobEvents(updates)
+		}
+	}()
+
+	return nil
+}
+
+func (sm *subscriptionManager) Stop() {
+	sm.wg.Wait()
+
+	// Remove all subscriptions and close corresponding channels.
+	func() {
+		sm.mu.Lock()
+		defer sm.mu.Unlock()
+
+		for subID, sub := range sm.subscriptions {
+			close(sub.Chan)
+			delete(sm.subscriptions, subID)
+		}
+	}()
+}
+
+func (sm *subscriptionManager) logStats(ctx context.Context, svcName string) {
+	// TODO: don't use mutex + struct internals externally like this
```

Comment:
I'm having a hard time understanding this TODO — can we fix it?

Reply:
Derp, leftover note to self that I forgot to clean up after I finished the refactor. Thanks